### PR TITLE
Enhancement/update documentation

### PR DESCRIPTION
**Pull request by cstrangedk**

### CODE REVIEW

File: contracts/ConstantsHolder.sol [hunk 1]
```diff
@@ -133,49 +131,71 @@ contract ConstantsHolder is Permissions {
     }
 
     /**
-     * Set new check time. This function only for tests.
-     * @param newCheckTime - new check time
+     * @dev Allows the Owner to set the new check time.
+     * This function only for tests.
```
Comment (gannakulikova) [inline contracts/ConstantsHolder.sol]: is only for tests?

File: contracts/Monitors.sol [hunk 2]
```diff
@@ -275,20 +341,29 @@ contract Monitors is Permissions {
         }
     }
 
+    /**
+     * @dev Returns the median using the Quicksort algorithm.
+     */
     function _median(uint[] memory values) private pure returns (uint) {
         if (values.length < 1) {
-            revert("Can't calculate _median of empty array");
+            revert("Cannot calculate _median of empty array");
```
Comment (gannakulikova) [inline contracts/Monitors.sol]: an empty array?

File: contracts/Nodes.sol [hunk 3]
```diff
@@ -31,7 +33,18 @@ import "./delegation/DelegationController.sol";
...
+ * @title Nodes
+ * @dev This contract contains all logic to manage SKALE Network nodes states,
+ * space availability, stake requirement checks, and exit functions.
+ * 
+ * Nodes may be in one of several states:
+ * 
+ * - Active:            Node is registered and is in network operation.
+ * - Leaving:           Node has begun exiting from the network.
+ * - Left:              Node has left the network.
+ * - In_Maintenance:    Node is temporarily offline.
```
Comment (gannakulikova) [inline contracts/Nodes.sol]: I believe we can describe 'In Maintenance' status as 'Node is temporarily offline or undergoing infrastructure maintenance'? What do you think?

File: contracts/SkaleDKG.sol [hunk 4]
```diff
@@ -77,22 +82,54 @@ contract SkaleDKG is Permissions, ISkaleDKG {
...
+
+    /**
+     * @dev Emitted when DKG is successful.
+     */
     event SuccessfulDKG(bytes32 indexed groupIndex);
+
+    /**
+     * @dev Emitted when a node is validated to be malicious.
+     */
     event BadGuy(uint nodeIndex);
```
Comment (gannakulikova) [inline contracts/SkaleDKG.sol:120]: is this a good wording?

File: contracts/SkaleDKG.sol [hunk 5]
```diff
@@ -77,22 +82,54 @@ contract SkaleDKG is Permissions, ISkaleDKG {
...
+
+    /**
+     * @dev Emitted when a complaint is sent.
+     */
     event ComplaintSent(bytes32 indexed groupIndex, uint indexed fromNodeIndex, uint indexed toNodeIndex);
+
+    /**
+     * @dev Emitted when a new node is rotated in to restart DKG.
+     */
     event NewGuy(uint nodeIndex);
```
Comment (gannakulikova) [inline contracts/SkaleDKG.sol:136]: is this a good wording?

File: contracts/Monitors.sol (hunk 2 above)
Comment (DimaStebaev) [reply contracts/Monitors.sol]: It's better to use `require` here.

File: contracts/SkaleToken.sol [hunk 6]
```diff
@@ -53,12 +54,14 @@ contract SkaleToken is ERC777, Permissions, ReentrancyGuard, IDelegatableToken {
...
-     * @param operatorData bytes extra information provided by the operator (if any)
-     * @return returns success of function call.
+     * @dev Allows SkaleManager to mint an amount of tokens and transfer minted
+     * tokens to a specified address.
+     * 
+     * Returns whether the operation is successful.
+     * 
+     * Requirements:
+     * 
+     * - Mint must not exceed the total supply.
```
Comment (DimaStebaev) [inline contracts/SkaleToken.sol:65]: Can be called only by owner or SkaleManager.

File: contracts/delegation/DelegationController.sol [hunk 7]
```diff
@@ -41,19 +41,22 @@ import "./PartialDifferences.sol";
...
- * - ACCEPTED: token delegations are accepted by a validator and are locked-by-delegation
- * - CANCELED: token holder cancels delegation proposal. Only allowed before the proposal is accepted by the validator
- * - REJECTED: token proposal expires at the UTC start of the next month
- * - DELEGATED: accepted delegations are delegated at the UTC start of the month
- * - UNDELEGATION_REQUESTED: token holder requests delegations to undelegate from the validator
- * - COMPLETED: undelegation request is completed at the end of the delegation period
+ * 
+ * - PROPOSED: token holder proposes tokens to delegate to a validator.
+ * - ACCEPTED: token delegations are accepted by a validator and are locked-by-delegation.
+ * - CANCELED: token holder cancels delegation proposal. Only allowed before the proposal is accepted by the .
```
Comment (DimaStebaev) [inline contracts/delegation/DelegationController.sol]: typo?
By the validator?

File: contracts/delegation/DelegationController.sol [hunk 8]
```diff
@@ -203,14 +206,23 @@ contract DelegationController is Permissions, ILocker {
         _;
     }
 
+    /**
+     * @dev Return and update validator's delegations.
```
Comment (DimaStebaev) [inline contracts/delegation/DelegationController.sol]: Maybe "update and return" is better? It first calculates value, stores it and then returns.

File: contracts/delegation/DelegationController.sol [hunk 9]
```diff
@@ -41,19 +41,22 @@ import "./PartialDifferences.sol";
  * @title Delegation Controller
  * @dev This contract performs all delegation functions including delegation
  * requests, undelegation, slashing, etc.
```
Comment (DimaStebaev) [inline contracts/delegation/DelegationController.sol]: Slashing is managed by `Punisher` not by `DelegationController`

File: contracts/delegation/DelegationController.sol [hunk 10]
```diff
@@ -428,19 +444,15 @@ contract DelegationController is Permissions, ILocker {
     }
 
     /**
-     * @dev Allows the Punisher to confiscate an `amount` of stake from
-     * `validatorId` by slashing. This slashes all delegations of the validator,
+     * @dev Allows Punisher contract to slash an `amount` of stake from
+     * a validator. This slashes all delegations of the validator,
```
Comment (DimaStebaev) [inline contracts/delegation/DelegationController.sol]: Not all but amount specified in `amount` parameter.

File: contracts/delegation/DelegationController.sol [hunk 11]
```diff
@@ -483,24 +514,32 @@ contract DelegationController is Permissions, ILocker {
         Permissions.initialize(contractsAddress);
     }
 
+    /**
+     * @dev Allows Nodes contract to get and update the amount delegated
+     * to validator.
```
Comment (DimaStebaev) [inline contracts/delegation/DelegationController.sol]: In particular month.

File: contracts/delegation/DelegationController.sol [hunk 12]
```diff
@@ -483,24 +514,32 @@ contract DelegationController is Permissions, ILocker {
...
+     */
     function getAndUpdateDelegatedToValidator(uint validatorId, uint month)
         public allow("Nodes") returns (uint)
     {
         return _delegatedToValidator[validatorId].getAndUpdateValue(month);
     }
 
+    /**
+     * @dev Process slashes.
```
Comment (DimaStebaev) [inline contracts/delegation/DelegationController.sol]: But not more than value in `limit` parameter.

File: contracts/delegation/DelegationController.sol [hunk 13]
```diff
@@ -864,6 +924,14 @@ contract DelegationController is Permissions, ILocker {
         );
     }
 
+    /**
+     * @dev Checks whether delegation to a validator is allowed.
+     * 
+     * Requirements:
+     * 
+     * - Delegator must not have reached the validator limit.
+     * - Delegation must be made in or after the first delegation month.
```
Comment (DimaStebaev) [inline contracts/delegation/DelegationController.sol:935]: Is not relevant anymore.

File: contracts/delegation/Distributor.sol [hunk 14]
```diff
@@ -77,10 +81,23 @@ contract Distributor is Permissions, IERC777Recipient {
...
+    /**
+     * @dev Return and update the amount of earned bounty from a validator.
+     */
     function getAndUpdateEarnedBountyAmount(uint validatorId) external returns (uint earned, uint endMonth) {
         return getAndUpdateEarnedBountyAmountOf(msg.sender, validatorId);
     }
 
+    /**
+     * @dev Allows msg.sender to withdraw earned bounty. Bounties are locked
+     * until launchTimestamp has passed.
```
Comment (DimaStebaev) [inline contracts/delegation/Distributor.sol]: Not `launchTimestamp` but `launchTimestamp` plus `BOUNTY_LOCKUP_MONTHS` value from `ConstantsHolder`. Currently 2 months.

File: contracts/delegation/Distributor.sol [hunk 15]
```diff
@@ -107,6 +124,16 @@ contract Distributor is Permissions, IERC777Recipient {
         );
     }
 
+    /**
+     * @dev Allows msg.sender to withdraw earned validator fees. Fees are locked
+     * until launchTimestamp has passed.
```
Comment (DimaStebaev) [inline contracts/delegation/Distributor.sol]: Not launchTimestamp but launchTimestamp plus BOUNTY_LOCKUP_MONTHS value from ConstantsHolder. Currently 2 months.

File: contracts/delegation/TokenLaunchLocker.sol [hunk 16]
```diff
@@ -67,12 +70,21 @@ contract TokenLaunchLocker is Permissions, ILocker {
...
+     * Emits a {Locked} event.
+     */
     function lock(address holder, uint amount) external allow("TokenLaunchManager") {
         _locked[holder] = _locked[holder].add(amount);
 
         emit Locked(holder, amount);
     }
 
+    /**
+     * @dev Allows DelegationController contract to add locked tokens to
```
Comment (DimaStebaev) [inline contracts/delegation/TokenLaunchLocker.sol]: Allows DelegationController contract to notify about new delegations.

File: contracts/delegation/TokenLaunchLocker.sol [hunk 17]
```diff
@@ -95,6 +107,10 @@ contract TokenLaunchLocker is Permissions, ILocker {
         }
     }
 
+    /**
+     * @dev Allows DelegationController contract to remove locked tokens from
```
Comment (DimaStebaev) [inline contracts/delegation/TokenLaunchLocker.sol]: Allows DelegationController contract to notify about new undelegation requests.

File: contracts/delegation/ValidatorService.sol [hunk 18]
```diff
@@ -214,17 +245,13 @@ contract ValidatorService is Permissions {
     }
 
     /**
-     * @dev Links a given node address.
-     *
+     * @dev Links a validator ID to a node address. Validator must present
```
Comment (DimaStebaev) [inline contracts/delegation/ValidatorService.sol]: I'd say "links a node address to validator ID"

File: docs/contract.hbs [hunk 19]
```diff
@@ -0,0 +1,67 @@
+{{~#*inline "typed-variable-array"~}}
```
Comment (DimaStebaev) [inline docs/contract.hbs:1]: Was this one committed by mistake?

File: contracts/Bounty.sol [hunk 20]
```diff
@@ -158,16 +168,25 @@ contract Bounty is Permissions {
...
+     * @dev Updates the number of nodes available for the current reward period.
+     */
     function _updateNodesPerRewardPeriod(ConstantsHolder constantsHolder, Nodes nodes) private {
         _nodesPerRewardPeriod = nodes.getNumberOnlineNodes();
         _nodesRemainingPerRewardPeriod = _nodesPerRewardPeriod;
         _rewardPeriodFinished = now.add(uint(constantsHolder.rewardPeriod()));
     }
 
+    /**
+     * @dev Returns the total available bounty pool a given stage.
```
Comment (payvint) [inline contracts/Bounty.sol]: "By" instead of "a"?

File: contracts/Decryption.sol [hunk 21]
```diff
@@ -21,13 +21,24 @@
 
 pragma solidity 0.6.10;
 
-
+/**
+ * @title Decryption
+ * @dev This contract performs encryption and decryption functions.
+ * Decrypt is used by SkaleDKG contract to decrypt secret key contribution to
+ * validate or invalidated complaints during the DKG procedure.
```
Comment (payvint) [inline contracts/Decryption.sol]: Does not clear for me, "to validate complaints" instead?

File: contracts/Decryption.sol [hunk 22]
```diff
@@ -21,13 +21,24 @@
...
 
+    /**
+     * @dev Returns an encrypted text given a secret and a key.
+     */
     function encrypt(uint256 secretNumber, bytes32 key) external pure returns (bytes32 ciphertext) {
         return bytes32(secretNumber) ^ key;
     }
 
+    /**
+     * @dev Returns a decrypted text given an encrypted text and a key.
```
Comment (payvint) [inline contracts/Decryption.sol]: Returns a secret(based on previous function doc)
`   /**
      * @dev Returns an encrypted text given a secret and a key.
      */`

File: contracts/Monitors.sol [hunk 23]
```diff
@@ -67,6 +74,9 @@ contract Monitors is Permissions {
         uint gasSpend
     );
 
+    /**
+     * @dev Emitted when a verdict is sent.
```
Comment (payvint) [inline contracts/Monitors.sol:78]: Does it mean that we need to change name of event "VerdictWasSent" to "VerdictIsSent"

File: contracts/Monitors.sol [hunk 24]
```diff
@@ -78,6 +88,9 @@ contract Monitors is Permissions {
         uint gasSpend
     );
 
+    /**
+     * @dev Emitted when metrics are calculated.
```
Comment (payvint) [inline contracts/Monitors.sol:92]: The same as previous comment

File: contracts/Monitors.sol [hunk 25]
```diff
@@ -86,6 +99,9 @@ contract Monitors is Permissions {
         uint gasSpend
     );
 
+    /**
+     * @dev Emitted when monitoring periods are set.
```
Comment (payvint) [inline contracts/Monitors.sol:103]: The same

File: contracts/Monitors.sol [hunk 26]
```diff
@@ -144,11 +168,23 @@ contract Monitors is Permissions {
         delete groupsForMonitors[monitorIndex];
     }
 
+    /**
+     * @dev Allows SkaleManager contract to remove a node from a monitor group.
```
Comment (payvint) [inline contracts/Monitors.sol]: Remove nodes from a monitoring list?
This function remove all nodes which given node monitoring to

File: contracts/Monitors.sol [hunk 27]
```diff
@@ -193,10 +233,16 @@ contract Monitors is Permissions {
...
 
+    /**
+     * @dev Allows SkaleManager contract to set a node's last bounty block.
+     */
     function setLastBountyBlock(uint nodeIndex) external allow("SkaleManager") {
         lastBountyBlocks[keccak256(abi.encodePacked(nodeIndex))] = block.number;
     }
 
+    /**
+     * @dev Returns node IPs that were last checked.
```
Comment (payvint) [inline contracts/Monitors.sol]: Returns node info(IPs, nodeIndex and time to send verdict) of nodes which are in monitoring list(MonitorIndex is monitoring this list)

File: contracts/Monitors.sol (hunk 2 above)
Comment (payvint) [reply contracts/Monitors.sol]: Agree

File: contracts/Monitors.sol [hunk 28]
```diff
@@ -240,14 +296,24 @@ contract Monitors is Permissions {
...
 
+    /**
+     * @dev Returns a count of metric data available for a node.
+     */
     function getLengthOfMetrics(bytes32 monitorIndex) public view returns (uint) {
         return verdicts[monitorIndex].length;
     }
 
+    /**
+     * @dev Allows SkaleManager contract to generate a monitoring group, using
```
Comment (payvint) [inline contracts/Monitors.sol]: Does not allow - private function

File: contracts/Nodes.sol [hunk 29]
```diff
@@ -154,10 +173,9 @@ contract Nodes is Permissions {
     }
 
     /**
-     * @dev adSpaceToFractionalNode - returns space to Fractional Node
-     * function could be run only be Schains
-     * @param nodeIndex - index of Node at array of Fractional Nodes
-     * @param space - space which should be returned
+     * @dev Allows Schains contract to occupy available capacity on a node.
```
Comment (payvint) [inline contracts/Nodes.sol]: Maybe to free or release space on node?

File: contracts/Nodes.sol [hunk 30]
```diff
@@ -333,6 +391,14 @@ contract Nodes is Permissions {
         return position.add(1).mul(msr) <= delegationsTotal;
     }
 
+    /**
+     * @dev Allows Node to set In_Maintenance status.
+     * 
+     * Requirements:
+     * 
+     * - Node must already be Active.
+     * - `msg.sender` must be owner of Node.
```
Comment (payvint) [inline contracts/Nodes.sol]: Not only the owner of Node - owner of Node, validator and SkaleManager owner

File: contracts/Nodes.sol [hunk 31]
```diff
@@ -345,8 +411,16 @@ contract Nodes is Permissions {
         nodes[nodeIndex].status = NodeStatus.In_Maintenance;
     }
 
+    /**
+     * @dev Allows Node to remove In_Maintenance status.
+     * 
+     * Requirements:
+     * 
+     * - Node must already be In Maintenance.
+     * - `msg.sender` must be owner of Node.
```
Comment (payvint) [inline contracts/Nodes.sol]: The same

File: contracts/Nodes.sol [hunk 32]
```diff
@@ -402,9 +479,7 @@ contract Nodes is Permissions {
     }
 
     /**
-     * @dev getNodePort - get Node's port
-     * @param nodeIndex - index of Node
-     * @return port
+     * @dev Returns the port of a given node.
```
Comment (payvint) [inline contracts/Nodes.sol:494]: Also "Node must exist"

**Review by payvint — COMMENTED** [review]
Need to discuss

File: contracts/Schains.sol [hunk 33]
```diff
@@ -175,6 +197,16 @@ contract Schains is Permissions {
         emit SchainDeleted(from, name, schainId);
     }
 
+    /**
+     * @dev Allows SkaleManager to delete a root owned schain.
```
Comment (payvint) [inline contracts/Schains.sol]: Not only root owned - Skale Manager Owner could remove every SKALE chain in the network

File: contracts/Schains.sol [hunk 34]
```diff
@@ -208,14 +240,25 @@ contract Schains is Permissions {
...
 
+    /**
+     * @dev Allows SkaleManager contract to restart schain creation by forming a
+     * new schain group.
+     * 
+     * Emits a {NodeAdded} event.
+     * 
+     * Requirements:
+     * 
+     * - Schain group must pass DKG.
```
Comment (payvint) [inline contracts/Schains.sol]: No - Schain group must be not completed!
This function must be called after:
- Previous DKG procedure was failed and did not find new node to add in group(15 nodes left)
- New node released in the network - so we could restart Schain creation


During this function new node become a part of the Schain group and DKG channel become opened

File: contracts/Schains.sol [hunk 35]
```diff
@@ -368,11 +403,9 @@ contract Schains is Permissions {
     }
 
     /**
-     * @dev _createGroupForSchain - creates Group for Schain
-     * @param schainName - name of Schain
-     * @param schainId - hash by name of Schain
-     * @param numberOfNodes - number of Nodes needed for this Schain
-     * @param partOfNode - divisor of given type of Schain
+     * @dev Allows SkaleManager to create group of nodes for Schain.
```
Comment (payvint) [inline contracts/Schains.sol]: Does not allow SkaleManager - it is private function

File: contracts/SchainsInternal.sol [hunk 36]
```diff
@@ -119,33 +118,26 @@ contract SchainsInternal is Permissions {
     }
 
     /**
-     * @dev setSchainIndex - adds Schain's hash to owner
-     * function could be run only by executor
-     * @param schainId - hash by Schain name
-     * @param from - Schain owner
+     * @dev Allows Schains contract to set schain index.
```
Comment (payvint) [inline contracts/SchainsInternal.sol]: index in owner list? or push schain list of chains by owner?
Does not clear what is schain index

File: contracts/SchainsInternal.sol [hunk 37]
```diff
@@ -219,20 +213,15 @@ contract SchainsInternal is Permissions {
     }
 
     /**
-     * @dev setException - sets a Node like exception
-     * function could be run only by executor
-     * @param schainId - Groups identifier
-     * @param nodeIndex - index of Node which would be notes like exception
+     * @dev Allows Schain contract to set a Node like exception for a given
```
Comment (payvint) [inline contracts/SchainsInternal.sol]: Allow Schains and NodeRotation

File: contracts/SchainsInternal.sol [hunk 38]
```diff
@@ -219,20 +213,15 @@ contract SchainsInternal is Permissions {
...
     function setException(bytes32 schainId, uint nodeIndex) external allowTwo("Schains", "NodeRotation") {
         _exceptionsForGroups[schainId][nodeIndex] = true;
     }
 
     /**
-     * @dev setNodeInGroup - adds Node to Group
-     * function could be run only by executor
-     * @param schainId - Groups
-     * @param nodeIndex - index of Node which would be added to the Group
+     * @dev Allows Schains contract to add node to an schain group.
```
Comment (payvint) [inline contracts/SchainsInternal.sol]: Schains and NodeRotation

File: contracts/SchainsInternal.sol [hunk 39]
```diff
@@ -263,92 +252,87 @@ contract SchainsInternal is Permissions {
...
      */
     function getSchains() external view returns (bytes32[] memory) {
         return schainsAtSystem;
     }
 
     /**
-     * @dev getSchainsPartOfNode - gets occupied space for given Schain
-     * @param schainId - hash by Schain name
-     * @return occupied space
+     * @dev Returns all occupied resources for an Schain
```
Comment (payvint) [inline contracts/SchainsInternal.sol]: I think not the all occupied - occupied resources on one node

File: contracts/SchainsInternal.sol [hunk 40]
```diff
@@ -263,92 +252,87 @@ contract SchainsInternal is Permissions {
...
 
+    /**
+     * @dev Returns schain name.
+     */
     function getSchainName(bytes32 schainId) external view returns (string memory) {
         return schains[schainId].name;
     }
 
+    /**
+     * @dev Returns active schains of a node.
```
Comment (payvint) [inline contracts/SchainsInternal.sol]: Returns last active schain of a node

File: contracts/SchainsInternal.sol [hunk 41]
```diff
@@ -408,6 +388,9 @@ contract SchainsInternal is Permissions {
         return schainsGroups[schainId].length;
     }
 
+    /**
+     * @dev Checks whether there are any schain nodes with free resources.
```
Comment (payvint) [inline contracts/SchainsInternal.sol]: Not schain nodes - any nodes with free resources which corresponds to this schain

File: contracts/SchainsInternal.sol [hunk 42]
```diff
@@ -441,10 +427,7 @@ contract SchainsInternal is Permissions {
     }
 
     /**
-     * @dev addSchainForNode - adds Schain hash to Node
-     * function could be run only by executor
-     * @param nodeIndex - index of Node
-     * @param schainId - hash by Schain name
+     * @dev Allows Schain contract to add schain to node.
```
Comment (payvint) [inline contracts/SchainsInternal.sol]: Schains and NodeRotation

File: contracts/SchainsInternal.sol [hunk 43]
```diff
@@ -470,10 +453,8 @@ contract SchainsInternal is Permissions {
     }
 
     /**
-     * @dev removesSchainForNode - clean given Node of Schain
-     * function could be run only by executor
-     * @param nodeIndex - index of Node
-     * @param schainIndex - index of Schain in schainsForNodes array by this Node
+     * @dev Allows Schains and SkaleDKG contracts to remove an schain from a
```
Comment (payvint) [inline contracts/SchainsInternal.sol]: And NodeRotation

File: contracts/SchainsInternal.sol [hunk 44]
```diff
@@ -495,19 +476,14 @@ contract SchainsInternal is Permissions {
...
     function getLengthOfSchainsForNode(uint nodeIndex) public view returns (uint) {
         return schainsForNodes[nodeIndex].length;
     }
 
     /**
-     * @dev findSchainAtSchainsForNode - finds index of Schain at schainsForNode array
-     * @param nodeIndex - index of Node at common array of Nodes
-     * @param schainId - hash of name of Schain
-     * @return index of Schain at schainsForNode array
+     * @dev Returns index of Schain in schainsForNode array.
```
Comment (payvint) [inline contracts/SchainsInternal.sol]: in list of schains on a given node

File: contracts/SkaleDKG.sol [hunk 45]
```diff
@@ -76,22 +81,54 @@ contract SkaleDKG is Permissions, ISkaleDKG {
...
+     */
     event BroadcastAndKeyShare(
         bytes32 indexed groupIndex,
         uint indexed fromNode,
         G2Operations.G2Point[] verificationVector,
         KeyShare[] secretKeyContribution
     );
 
+    /**
+     * @dev Emitted when all group data is received.
```
Comment (payvint) [inline contracts/SkaleDKG.sol]: ... by node

File: contracts/SkaleDKG.sol [hunk 46]
```diff
@@ -76,22 +81,54 @@ contract SkaleDKG is Permissions, ISkaleDKG {
...
+    /**
+     * @dev Emitted when a complaint is sent.
+     */
     event ComplaintSent(bytes32 indexed groupIndex, uint indexed fromNodeIndex, uint indexed toNodeIndex);
+
+    /**
+     * @dev Emitted when a new node is rotated in to restart DKG.
+     */
     event NewGuy(uint nodeIndex);
     event ComplaintError(string error);
```
Comment (payvint) [inline contracts/SkaleDKG.sol:141]: Emitted when an incorrect complaint is sent

File: contracts/SkaleDKG.sol [hunk 47]
```diff
@@ -137,6 +190,17 @@ contract SkaleDKG is Permissions, ISkaleDKG {
...
+    /**
+     * @dev Broadcasts verification vector and secret key contribution to all
+     * other nodes in the group.
+     *
+     * Emits BroadcastAndKeyShare event.
+     *
+     * Requirements:
+     *
+     * - msg.sender must have an associated node.
+     * - VerificationVector must not be empty.
```
Comment (payvint) [inline contracts/SkaleDKG.sol]: verificationVector must be a certain length
secretKeyContribution length must be as number of nodes in this group

File: contracts/SkaleDKG.sol [hunk 48]
```diff
@@ -172,6 +236,16 @@ contract SkaleDKG is Permissions, ISkaleDKG {
         );
     }
 
+    /**
+     * @dev Creates a complaint from a node (accuser) to a given node.
+     * The accusing node must broadcast additional parameters within 1800 blocks.
+     *
+     * Emits BroadcastAndKeyShare event.
```
Comment (payvint) [inline contracts/SkaleDKG.sol]: ComplaintSent - successful
ComplaintError - unsuccessful

File: contracts/SkaleDKG.sol [hunk 49]
```diff
@@ -280,11 +354,14 @@ contract SkaleDKG is Permissions, ISkaleDKG {
...
+     * @dev Checks whether channel is opened.
+     */
     function isChannelOpened(bytes32 groupIndex) external override view returns (bool) {
         return channels[groupIndex].active;
     }
 
-    function isLastDKGSuccesful(bytes32 groupIndex) external override view returns (bool) {
+    function isLastDKGSuccessful(bytes32 groupIndex) external override view returns (bool) {
         return channels[groupIndex].startedBlockTimestamp <= lastSuccesfulDKG[groupIndex];
     }
```
Comment (payvint) [inline contracts/SkaleDKG.sol:421]: Checks whether broadcast is possible

File: contracts/SkaleDKG.sol [hunk 50]
```diff
@@ -369,6 +455,9 @@ contract SkaleDKG is Permissions, ISkaleDKG {
         return channels[groupIndex].n == dkgProcess[groupIndex].numberOfBroadcasted;
     }
 
+    /**
+     * @dev Checks whether all data has been received.
```
Comment (payvint) [inline contracts/SkaleDKG.sol]: ... by node

**Review by payvint — COMMENTED** [review]
Need to discuss

File: contracts/SkaleVerifier.sol [hunk 51]
```diff
@@ -27,10 +27,24 @@ import "./SchainsInternal.sol";
...
+
+    /**
+    * @dev Verifies a BLS signature.
+    * 
+    * Requirements:
+    * 
+    * - Signature is in G1.
+    * - Hash is in G1.
+    * - G2.one in G2.
+    * - Public Key in G1.
```
Comment (payvint) [inline contracts/SkaleVerifier.sol]: Public Key in G2

File: contracts/delegation/DelegationController.sol [hunk 52]
```diff
@@ -203,14 +206,23 @@ contract DelegationController is Permissions, ILocker {
...
 
+    /**
+     * @dev Return and update the amount delegated.
+     */
     function getAndUpdateDelegatedAmount(address holder) external returns (uint) {
         return _getAndUpdateDelegatedByHolder(holder);
     }
 
+    /**
+     * @dev Return and update the effective amount delegated (minus slash).
```
Comment (payvint) [inline contracts/delegation/DelegationController.sol]: Maybe it better to add in given month param

File: contracts/delegation/DelegationController.sol [hunk 53]
```diff
@@ -451,30 +463,49 @@ contract DelegationController is Permissions, ILocker {
...
+     * amount delegated (minus slash) to a validator.
+     */
     function getAndUpdateEffectiveDelegatedToValidator(uint validatorId, uint month)
         external allow("Distributor") returns (uint)
     {
         return _effectiveDelegatedToValidator[validatorId].getAndUpdateValueInSequence(month);
     }
 
+    /**
+     * @dev Return and update the amount delegated to a validator.
```
Comment (payvint) [inline contracts/delegation/DelegationController.sol]: In current month

File: contracts/delegation/DelegationController.sol [hunk 54]
```diff
@@ -451,30 +463,49 @@ contract DelegationController is Permissions, ILocker {
         _slashes.push(SlashingEvent({reducingCoefficient: coefficient, validatorId: validatorId, month: currentMonth}));
     }
 
+    /**
+     * @dev Allows Distributor contract to return and update the effective 
+     * amount delegated (minus slash) to a validator.
```
Comment (payvint) [inline contracts/delegation/DelegationController.sol]: in a given month

File: contracts/delegation/DelegationPeriodManager.sol [hunk 55]
```diff
@@ -43,13 +43,10 @@ contract DelegationPeriodManager is Permissions {
...
     /**
-     * @dev Creates a new available delegation period and return in the network.
-     * Only the owner may set new delegation period and returns in the network.
-     *
-     * Emits a DelegationPeriodWasSet event.
-     *
-     * @param monthsCount uint delegation duration in months
-     * @param stakeMultiplier uint return for delegation
+     * @dev Allows the Owner to create a new available delegation period and
+     * return in the network.
```
Comment (payvint) [inline contracts/delegation/DelegationPeriodManager.sol]: Does not clear - return WHAT in the network?
Maybe without return in the network

File: contracts/delegation/Distributor.sol [hunk 56]
```diff
@@ -152,6 +179,9 @@ contract Distributor is Permissions, IERC777Recipient {
         _distributeBounty(amount, validatorId);
     }
 
+    /**
+     * @dev Return the amount of earned validator fees.
```
Comment (payvint) [inline contracts/delegation/Distributor.sol]: By msg.sender?

File: contracts/delegation/Distributor.sol [hunk 57]
```diff
@@ -198,6 +231,9 @@ contract Distributor is Permissions, IERC777Recipient {
         }
     }
 
+    /**
+     * @dev Return the amount of earned fees.
```
Comment (payvint) [inline contracts/delegation/Distributor.sol]: By validator ID?

File: contracts/delegation/Punisher.sol [hunk 58]
```diff
@@ -76,11 +76,12 @@ contract Punisher is Permissions, ILocker {
 
     /**
      * @dev Allows the Owner to forgive a slashing condition.
```
Comment (payvint) [inline contracts/delegation/Punisher.sol]: Admin instead of Owner

File: contracts/delegation/TokenLaunchLocker.sol [hunk 59]
```diff
@@ -67,12 +70,21 @@ contract TokenLaunchLocker is Permissions, ILocker {
     // delegationId => tokens
     mapping (uint => uint) private _delegationAmount;
 
+    /**
+     * @dev Allows TokenLaunchManager contract to lock an amount of tokens.
```
Comment (payvint) [inline contracts/delegation/TokenLaunchLocker.sol]: in holder wallet?

File: contracts/delegation/TokenLaunchLocker.sol [hunk 60]
```diff
@@ -148,48 +170,74 @@ contract TokenLaunchLocker is Permissions, ILocker {
...
 
+    /**
+     * @dev Returns and updates the current delegated amount.
+     */
     function _getAndUpdateDelegatedAmount(address holder, uint currentMonth) private returns (uint) {
         return _delegatedAmount[holder].getAndUpdateValue(currentMonth);
     }
 
+    /**
+     * @dev Adds a delegated amount to the current month.
```
Comment (payvint) [inline contracts/delegation/TokenLaunchLocker.sol]: to the given month instead of current

File: contracts/delegation/TokenLaunchLocker.sol [hunk 61]
```diff
@@ -148,48 +170,74 @@ contract TokenLaunchLocker is Permissions, ILocker {
...
+     */
     function _addToDelegatedAmount(address holder, uint amount, uint month) private {
         _delegatedAmount[holder].addToValue(amount, month);
     }
 
+    /**
+     * @dev Removes a delegated amount from the current month.
```
Comment (payvint) [inline contracts/delegation/TokenLaunchLocker.sol]: to the given month instead of current

File: contracts/delegation/TokenLaunchLocker.sol [hunk 62]
```diff
@@ -148,48 +170,74 @@ contract TokenLaunchLocker is Permissions, ILocker {
...
+     */
     function _removeFromDelegatedAmount(address holder, uint amount, uint month) private {
         _delegatedAmount[holder].subtractFromValue(amount, month);
     }
 
+    /**
+     * @dev Adds the amount to the total delegated for the current month.
```
Comment (payvint) [inline contracts/delegation/TokenLaunchLocker.sol]: to the given month instead of current

File: contracts/delegation/TokenLaunchManager.sol [hunk 63]
```diff
@@ -112,34 +109,34 @@ contract TokenLaunchManager is Permissions, IERC777Recipient {
     }
 
     /**
-     * @dev Allow withdrawals and disallow approvals changes
-     *
+     * @dev Allows the Seller to complete the token launch process.
+     * This then allows participants to retrieve tokens and this disallows
```
Comment (payvint) [inline contracts/delegation/TokenLaunchManager.sol]: This allows - remove then?

File: contracts/delegation/ValidatorService.sol [hunk 64]
```diff
@@ -173,22 +197,20 @@ contract ValidatorService is Permissions {
...
+     * 
      * Requirements:
-     *
-     * - new address must not be null
-     * - new address must not be already registered as a validator
-     *
-     * @param newValidatorAddress address
+     * 
+     * - New address must not be null.
+     * - New address must not be already registered as a validator.
```
Comment (payvint) [inline contracts/delegation/ValidatorService.sol:214]: Also `msg.sender` must be a validator

**Review by payvint — COMMENTED** [review]
Finally!!!!!!!!!
Still need to discuss

P.s. It was a long journey

File: contracts/Schains.sol (hunk 33 above)
Comment (cstrangedk) [reply contracts/Schains.sol]: then what is difference between deleteSchain and deleteSchainByRoot?

Comment (payvint) [reply contracts/Schains.sol]: So deleteSchain - is called by the owner of Schains
deleteSchainByRoot - by the Admin

File: contracts/SkaleDKG.sol (hunk 4 above)
Comment (cstrangedk) [reply contracts/SkaleDKG.sol:120]: perhaps `verification` is better

File: docs/contract.hbs (hunk 19 above)
Comment (cstrangedk) [reply docs/contract.hbs:1]: This is intentional as to have our template for docgen available

File: contracts/SkaleDKG.sol (hunk 5 above)
Comment (cstrangedk) [reply contracts/SkaleDKG.sol:136]: 1. DKG fails, and BadGuy and FailedDKG are emitted
2. a free node is requested
3. bad node is rotated out, new free node is rotated in
DKG is restarted

Suggest to simply to ...when a new node is rotated in.

**Review by DimaStebaev — APPROVED** [review]

**Review by yavrsky — APPROVED** [review]

**Pull request merged**